// docpad-plugin-partials/es5/partials.plugin.js
// Generated by CoffeeScript 1.12.7
var extend = function(child, parent) { for (var key in parent) { if (hasProp.call(parent, key)) child[key] = parent[key]; } function ctor() { this.constructor = child; } ctor.prototype = parent.prototype; child.prototype = new ctor(); child.__super__ = parent.prototype; return child; },
  hasProp = {}.hasOwnProperty,
  slice = [].slice;

module.exports = function(BasePlugin) {
  var PartialsPlugin, Task, TaskGroup, extendr, pathUtil, ref, util;
  extendr = require('extendr');
  ref = require('taskgroup'), Task = ref.Task, TaskGroup = ref.TaskGroup;
  pathUtil = require('path');
  util = require('util');
  return PartialsPlugin = (function(superClass) {
    extend(PartialsPlugin, superClass);

    PartialsPlugin.prototype.name = 'partials';

    PartialsPlugin.prototype.config = {
      partialPaths: ['partials'],
      collectionName: 'partials',
      performanceFirst: false
    };

    PartialsPlugin.prototype.locale = {
      addingPartial: "Adding partial: %s",
      partialNotFound: "The partial \"%s\" was not found, as such it will not be rendered.",
      renderPartial: "Rendering partial: %s",
      renderedPartial: "Rendered partial: %s",
      renderPartialFailed: "Rendering partial failed: %s. The error follows:"
    };

    PartialsPlugin.prototype.foundPartials = null;

    PartialsPlugin.prototype.partialsCache = null;

    function PartialsPlugin() {
      PartialsPlugin.__super__.constructor.apply(this, arguments);
      this.prepareConfig();
      this.partialsCache = {};
      this.foundPartials = {};
    }

    PartialsPlugin.prototype.setConfig = function() {
      PartialsPlugin.__super__.setConfig.apply(this, arguments);
      this.prepareConfig();
      return this;
    };

    PartialsPlugin.prototype.prepareConfig = function() {
      var config, docpadConfig;
      docpadConfig = this.docpad.getConfig();
      config = this.getConfig();
      config.partialPaths = config.partialsPath || config.partialPaths;
      if (!util.isArray(config.partialPaths)) {
        config.partialPaths = [config.partialPaths];
      }
      config.partialPaths.forEach(function(partialPath, index) {
        config.partialPaths[index] = pathUtil.resolve(docpadConfig.srcPath, partialPath);
      });
      return this;
    };

    PartialsPlugin.prototype.populateCollections = function(opts, next) {
      var config, docpad, processedCount;
      config = this.config;
      docpad = this.docpad;
      processedCount = 0;
      config.partialPaths.forEach(function(partialPath) {
        docpad.parseDocumentDirectory({
          path: partialPath
        }, function(err, results) {
          if (err || processedCount === (config.partialPaths.length - 1)) {
            next(err);
          }
          processedCount++;
        });
      });
      return this;
    };

    PartialsPlugin.prototype.extendCollections = function(opts) {
      var config, database, docpad, locale;
      config = this.getConfig();
      docpad = this.docpad;
      locale = this.locale;
      database = docpad.getDatabase();
      docpad.setCollection(config.collectionName, database.createLiveChildCollection().setQuery('isPartial', {
        $or: {
          isPartial: true,
          fullPath: {
            $startsWith: config.partialPaths
          }
        }
      }).on('add', function(model) {
        docpad.log('debug', util.format(locale.addingPartial, model.getFilePath()));
        return model.setDefaults({
          isPartial: true,
          render: false,
          write: false
        });
      }));
      return this;
    };

    PartialsPlugin.prototype.renderPartial = function(partial, next) {
      var cacheable, docpad, locale, partialsCache, ref1, ref2, result;
      docpad = this.docpad;
      locale = this.locale;
      partialsCache = this.partialsCache;
      result = null;
      cacheable = (ref1 = partial.document.getMeta().get('cacheable')) != null ? ref1 : false;
      if (cacheable === true) {
        result = (ref2 = partialsCache[partial.cacheId]) != null ? ref2 : null;
      }
      if (result != null) {
        return next(null, result);
      }
      docpad.renderDocument(partial.document, {
        templateData: partial.data
      }, function(err, result, document) {
        if (err) {
          return next(err);
        }
        if (cacheable === true) {
          partialsCache[partial.cacheId] = result;
        }
        return next(null, result);
      });
      return this;
    };

    PartialsPlugin.prototype.extendTemplateData = function(arg) {
      var docpad, locale, me, templateData;
      templateData = arg.templateData;
      me = this;
      docpad = this.docpad;
      locale = this.locale;
      templateData.partial = function() {
        var collection, config, err, file, i, len, message, obj, objs, partial, partialName, ref1;
        partialName = arguments[0], objs = 2 <= arguments.length ? slice.call(arguments, 1) : [];
        config = me.getConfig();
        if (typeof this.referencesOthers === "function") {
          this.referencesOthers();
        }
        file = this.documentModel;
        partial = {};
        collection = docpad.getCollection('partials');
        config.partialPaths.forEach(function(partialPath) {
          var partialFuzzyPath;
          partialFuzzyPath = pathUtil.join(partialPath, partialName);
          if (partial.document == null) {
            partial.document = collection.fuzzyFindOne(partialFuzzyPath);
          }
        });
        if (!partial.document) {
          message = util.format(locale.partialNotFound, partialName);
          err = new Error(message);
          if (partial.err == null) {
            partial.err = err;
          }
          return message;
        }
        partial.data = {};
        if (config.performanceFirst === false) {
          if ((ref1 = objs[0]) !== false && ref1 !== this) {
            objs.unshift(this);
          }
        }
        for (i = 0, len = objs.length; i < len; i++) {
          obj = objs[i];
          if (obj && (obj !== true)) {
            extendr.extend(partial.data, obj);
          }
        }
        partial.path = partial.document.getFilePath();
        partial.cacheId = partial.document.id;
        partial.id = Math.random();
        partial.container = '[partial:' + partial.id + ']';
        if (me.foundPartials[partial.id]) {
          return partial.container;
        }
        me.foundPartials[partial.id] = partial;
        partial.task = new Task("renderPartial: " + partial.path, function(complete) {
          return me.renderPartial(partial, function(err, result) {
            var ref2, ref3, ref4;
            if (partial.err == null) {
              partial.err = err;
            }
            partial.result = (ref2 = (ref3 = (ref4 = partial.err) != null ? ref4.toString() : void 0) != null ? ref3 : result) != null ? ref2 : '???';
            return complete(partial.err);
          });
        });
        return partial.container;
      };
      return this;
    };

    PartialsPlugin.prototype.renderDocument = function(opts, next) {
      var file, filePath, me, partialContainerRegex, partialContainers, tasks, templateData;
      templateData = opts.templateData, file = opts.file;
      partialContainerRegex = /\[partial:([^\]]+)\]/g;
      partialContainers = (opts.content || '').match(partialContainerRegex) || [];
      if (partialContainers.length === 0) {
        return next();
      }
      filePath = file.getFilePath();
      me = this;
      tasks = new TaskGroup("Partials for " + filePath, {
        concurrency: 0
      }).done(function(err) {
        opts.content = opts.content.replace(partialContainerRegex, function(match, partialId) {
          var partial;
          partial = me.foundPartials[partialId];
          return partial.result;
        });
        return next(err);
      });
      partialContainers.forEach(function(partialContainer) {
        var partial, partialId;
        partialId = partialContainer.replace(partialContainerRegex, '$1');
        partial = me.foundPartials[partialId];
        if (partial.task) {
          return tasks.addTask(partial.task);
        }
      });
      tasks.run();
      return this;
    };

    PartialsPlugin.prototype.generateAfter = function() {
      this.foundPartials = {};
      return this.partialsCache = {};
    };

    return PartialsPlugin;

  })(BasePlugin);
};
